README.org
* =orgability=
Orgability is a reading list manager with offline access support. Orgability
lets you to easily create topics between reading list entries and =org-brain=
entries by adding topics to reading entries.

** Installation
This package requires Emacs 25.1 or higher.

*** MELPA
This package is not yet available on MELPA.

*** Manual
Install =org-mode=, =org-cliplink=, =org-board= and =org-brain= and then require
this package in your =init.el= file:

#+BEGIN_SRC emacs-lisp
  (require 'orgability)
#+END_SRC

** Usage
*** Relations
There are several relations that one can create between different entries.
- =topic <-> resource= relation, where =org-brain= entry is a topic for reading
  list entry and reading list entry is a resource for =org-brain= entry.

*** Commands
- =orgability-clip=: Stores an URL from the clipboard to the =orgability-file=.
  - The title is fetched using =org-cliplink=.
  - When =orgability-auto-archive= is non-nil the URL is archived using
    =org-board= for an offline access.
- =orgability-store-url=: The same as =orgability-clip=, but asks for an URL.
- =orgability-open=: Opens archive using =org-board= in the
  =org-board-default-browser=.
- =orgability-add-topic=: Asks for a =org-brain= entry and creates a =topic <->
  resource= relation between reading list entry at the point and selected
  =org-brain= entry. Reading list entry is resource of =org-brain= entry and
  =org-brain= entry is a topic of reading list entry.
- =orgability-delete-topic=: Removes =topic <-> resource= relation between
  =org-brain= entry and reading list entry in a two-way fashion. That's why you
  should not manually remove topics from =orgability-file=.

*** Viewing reading list
Currently, =orgability= doesn't expose any special functionality to view all
your reading list, but you can use =org-agenda= for this purpose. There is a
function =orgability-agenda-format-topics= that can be used in
=org-agenda-prefix-format= to insert topics.

#+BEGIN_SRC emacs-lisp
  ("r" "Reading List" tags-todo "READING"
   ((org-agenda-overriding-header "Reading List")
    (org-agenda-remove-tags t)
    (org-agenda-prefix-format '((tags . "%(orgability-agenda-format-topics)")))
    (org-agenda-sorting-strategy '(todo-state-down effort-up category-keep))))
#+END_SRC

*** Configurations
- =orgability-file=: File with reading list entries.
- =orgability-add-to-top=: If non-nil, entries are added to the top of the
  =orgability-file=.
- =orgability-todo-keyword=: If non-nil, entries are added with a selected todo
  keyword.
- =orgability-active-timestamp=: If non-nil, ADDED timestamp is active and
  inactive otherwise.
- =orgability-auto-archive=: If non-nil, use =org-board= to archive reading list
  for offline access when adding it.
- =orgability-use-relative-archive-url=: If non-nil, use relative links to
  archived reading list entry. Useful when using several computers with
  different =$HOME= directories.
- =orgability-auto-agenda-redo=: If non-nil, refresh =org-agenda= when adding
  reading list entry.
- =orgability-title=: Title for a new =orgability-file=.
- =orgability-category=: Category for a new =orgability-file=.
- =orgability-file-tags=: File tags for a new =orgability-file=.
- =orgability-extract-http-title-f=: Function to extract title from http(s) URL.
  By default it's =org-cliplink-retrieve-title-synchronously= from a
  =org-cliplink= package. This variable allows you to choose your own
  implementation.
- =orgability-agenda-topics-column=: Width of topics block in
  =org-agenda=.

** Running tests
Ideally, you should use [[http://cask.readthedocs.io/en/latest/][Cask]] for running tests. On macOS one can install it
using following brew:

#+BEGIN_SRC bash
$ brew install cask [--ignore-dependencies]
#+END_SRC

After that you need to setup cask:

#+BEGIN_SRC bash
$ cask install
#+END_SRC

And now you are ready to run tests:

#+BEGIN_SRC bash
$ cask exec ert-runner
#+END_SRC

** Special thanks
Special thanks to creators and maintainers of =org-mode=, =org-cliplink=,
=org-brain= and =org-board=. And to all the Emacs community.

** Contributing
Don't hesitate to ask questions, to report bugs or suggest new features. And
most importantly, PRs are welcome!


CHANGELOG.org
* Unreleased
- New =org-agenda= helper for inserting relations list into agenda -
  =orgability-agenda-format-relations=.
- Fix =orgability-add-relation= when used from agenda.
- Automatically redo =org-agenda= after relations manipulations.
- Generalize drawers helpers (they can be used outside of orgability and
  probably will be moved to vulpea at some point of time).
- *Breaking change:* rename relations to topics.
- Introduce tests.
- Ask user for a title if it can't be retrieved from the web page itself.

* ~v0.1.0~
Initial release.


test/test-helper.el
(require 'cl)
(require 'el-mock)
(require 'undercover)
(require 'subr-x)

(setq user-home-directory (concat (getenv "HOME") "/"))

(undercover "*.el")

(add-to-list 'load-path ".")
(load "orgability.el")

(defmacro with-one-header-and-no-drawer (&rest body)
  `(with-temp-buffer
     (org-mode)
     (insert "* Some heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 8366A66A-2DE6-401B-AF7F-0C03C33EA3BB\n")
     (insert ":END:\n")
     ,@body))

(defmacro with-one-header-and-empty-drawer (&rest body)
  `(with-temp-buffer
     (org-mode)
     (insert "* Some heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 8366A66A-2DE6-401B-AF7F-0C03C33EA3BB\n")
     (insert ":END:\n")
     (insert ":TOPICS:\n")
     (insert ":END:\n")
     ,@body))

(defmacro with-one-header-and-drawer (&rest body)
  `(with-temp-buffer
     (org-mode)
     (insert "* Some heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 8366A66A-2DE6-401B-AF7F-0C03C33EA3BB\n")
     (insert ":END:\n")
     (insert ":TOPICS:\n")
     (insert "- topic1\n")
     (insert "- topic2\n")
     (insert ":END:\n")
     ,@body))

(defmacro with-two-headers-and-no-drawer (&rest body)
  `(with-temp-buffer
     (org-mode)
     (insert "* Some heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 8366A66A-2DE6-401B-AF7F-0C03C33EA3BB\n")
     (insert ":END:\n")
     (insert "* Main heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 6845FDEF-EAB9-423D-8792-8E736DCD414D\n")
     (insert ":END:\n")
     ,@body))

(defmacro with-two-headers-and-empty-drawer (&rest body)
  `(with-temp-buffer
     (org-mode)
     (insert "* Some heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 8366A66A-2DE6-401B-AF7F-0C03C33EA3BB\n")
     (insert ":END:\n")
     (insert "* Main heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 6845FDEF-EAB9-423D-8792-8E736DCD414D\n")
     (insert ":END:\n")
     (insert ":TOPICS:\n")
     (insert ":END:\n")
     ,@body))

(defmacro with-two-headers-and-drawer (&rest body)
  `(with-temp-buffer
     (org-mode)
     (insert "* Some heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 8366A66A-2DE6-401B-AF7F-0C03C33EA3BB\n")
     (insert ":END:\n")
     (insert "* Main heading\n")
     (insert ":PROPERTIES:\n")
     (insert ":ID: 6845FDEF-EAB9-423D-8792-8E736DCD414D\n")
     (insert ":END:\n")
     (insert ":TOPICS:\n")
     (insert "- topic1\n")
     (insert "- topic2\n")
     (insert ":END:\n")
     ,@body))


test/orgability-drawer-test.el
(ert-deftest orgability--drawer-block-test ()
  "Test that `orgability--drawer-block' works as expected."
  ;; no drawer
  (should
   (equal
    nil
    (with-one-header-and-no-drawer
     (orgability--drawer-block "topics"))))
  ;; empty drawer
  (should
   (equal
    (cons 77 86)
    (with-one-header-and-empty-drawer
     (orgability--drawer-block "topics"))))
  ;; empty drawer + inside
  (should
   (equal
    (cons 86 86)
    (with-one-header-and-empty-drawer
     (orgability--drawer-block "topics" nil t))))
  ;; drawer
  (should
   (equal
    (cons 77 104)
    (with-one-header-and-drawer
     (orgability--drawer-block "topics"))))
  ;; drawer + inside
  (should
   (equal
    (cons 86 104)
    (with-one-header-and-drawer
     (orgability--drawer-block "topics" nil t))))
  ;; no drawer + force == empty drawer
  (should
   (equal
    (with-one-header-and-empty-drawer
     (orgability--drawer-block "topics"))
    (with-one-header-and-no-drawer
     (orgability--drawer-block "topics" t))))
  ;; no drawer + force + inside == empty drawer + inside
  (should
   (equal
    (with-one-header-and-empty-drawer
     (orgability--drawer-block "topics" nil t))
    (with-one-header-and-no-drawer
     (orgability--drawer-block "topics" t t)))))


orgability-brain.el
;;; orgability-brain.el --- reading list manager in `org-mode'

;; Copyright (c) 2018 Boris Buliga

;; Author: Boris Buliga <boris@d12frosted.io>
;; Maintainer: Boris Buliga <boris@d12frosted.io>
;; Created: 25 Apr 2018

;; Keywords: org-mode
;; Homepage: https://github.com/d12frosted/orgability

;; Package-Version: 0.2.0
;; Package-Requires: ((org "9.1.0") (org-brain "0.5"))

;; This file is not part of GNU Emacs.
;;; License: GPLv3

;;; Commentary:
;;

;;; Code:
;;

(require 'org-brain)

(defun orgability-brain-choose-entry ()
  "PROMPT for an brain entry and return it."
  (org-brain-choose-entry
   "Entry: " (append (org-brain-files t)
                     (org-brain-headline-entries))))

(defun orgability-brain-add-resource (orgability-id brain-entry)
  "Add ORGABILITY-ID as resource of BRAIN-ENTRY."
  (org-brain-add-resource (concat "id:" orgability-id)
                          (org-entry-get nil "ITEM")
                          nil
                          brain-entry))

(defun orgability-brain-delete-resource (orgability-id brain-id)
  "Delete ORGABILITY-ID from BRAIN-ID resources."
  (when (null orgability-id)
    (error "orgability-id can't be null"))
  (when (null brain-id)
    (error "brain-id can't be null"))
  (let ((entry (or (org-brain-entry-from-id brain-id)
                   (org-brain-path-entry-name brain-id))))
    (if (org-brain-filep entry)
        (with-current-buffer (find-file-noselect (org-brain-entry-path entry))
          (goto-char (point-min))
          (or (re-search-forward (concat "^\\(" org-outline-regexp "\\)") nil t)
              (goto-char (point-max)))
          (when (re-search-backward org-brain-resources-start-re nil t)
            (orgability--remove-till (concat "^.*" orgability-id ".*$") ":END:")))
      (org-with-point-at (org-brain-entry-marker entry)
        (goto-char (cdr (org-get-property-block)))
        (forward-line 1)
        (when (looking-at org-brain-resources-start-re)
          (orgability--remove-till (concat "^.*" orgability-id ".*$") ":END:"))))))

(defun orgability-brain-get-link (brain-entry)
  "Get link to BRAIN-ENTRY."
  (if (org-brain-filep brain-entry)
      (concat "file://" (org-brain-entry-path brain-entry))
    (concat "id:" (org-brain-entry-identifier brain-entry))))

(provide 'orgability-brain)

;;; orgability-brain.el ends here


orgability-utils.el
;;; orgability-utils.el --- reading list manager in `org-mode'

;; Copyright (c) 2018 Boris Buliga

;; Author: Boris Buliga <boris@d12frosted.io>
;; Maintainer: Boris Buliga <boris@d12frosted.io>
;; Created: 25 Apr 2018

;; Keywords: org-mode
;; Homepage: https://github.com/d12frosted/orgability

;; Package-Version: 0.2.0
;; Package-Requires: ((org "9.1.0"))

;; This file is not part of GNU Emacs.
;;; License: GPLv3

;;; Commentary:
;;

;;; Code:
;;

(require 'org)

(defmacro orgability--with-entry (&rest body)
  "Move to buffer and point of current entry for the duration of BODY."
  `(cond ((eq major-mode 'org-mode)
          (org-with-point-at (point) ,@body))
         ((eq major-mode 'org-agenda-mode)
          (org-agenda-with-point-at-orig-entry nil ,@body))))

(defun orgability--remove-till (itemr stopr)
  "Remove lines matching ITEMR until STOPR is found."
  (while (not (looking-at stopr))
    (if (looking-at itemr)
        (kill-whole-line)
      (forward-line 1)
      (beginning-of-line))))

(defun orgability--unwrap-link (link)
  "Get the link without the type."
  (funcall (orgability--compose
            (apply-partially #'string-remove-prefix "id:")
            (apply-partially #'string-remove-prefix "file:"))
           link))

(defun orgability--lookup-key (key-map function)
  "Returns key binding for a FUNCTION in a KEY-MAP."
  (string (car (rassoc function (cdr key-map)))))

(defun orgability--compose (&rest funs)
  "Return function composed of FUNS."
  (lexical-let ((lex-funs funs))
    (lambda (&rest args)
      (reduce 'funcall (butlast lex-funs)
              :from-end t
              :initial-value (apply (car (last lex-funs)) args)))))

(provide 'orgability-utils)

;;; orgability-utils.el ends here


orgability-drawer.el
;;; orgability-drawer.el --- reading list manager in `org-mode'

;; Copyright (c) 2018 Boris Buliga

;; Author: Boris Buliga <boris@d12frosted.io>
;; Maintainer: Boris Buliga <boris@d12frosted.io>
;; Created: 25 Apr 2018

;; Keywords: org-mode
;; Homepage: https://github.com/d12frosted/orgability

;; Package-Version: 0.2.0
;; Package-Requires: ((org "9.1.0"))

;; This file is not part of GNU Emacs.
;;; License: GPLv3

;;; Commentary:
;;

;;; Code:
;;

(require 'org)
(require 'orgability-utils)

(defvar orgability--drawer-list-prefix "- ")

(defun orgability--drawer-block (name &optional force inside)
  "Return the (beg . end) range of the body of the NAME drawer.
If the drawer does not exist, create it if FORCE is non-nil, or
return nil."
  (setq name (upcase name))
  (orgability--with-entry
   (unless (org-before-first-heading-p)
     (org-back-to-heading t)
     (end-of-line)
     (let* ((drawer-beg-regexp (concat "^[ \t]*:" name ":[ \t]*$"))
            (drawer-end-regexp "^[ \t]*:END:[ \t]*$")
            (bound (save-excursion
                     (if (search-forward-regexp org-heading-regexp nil t)
                         (line-beginning-position)
                       (buffer-end 1))))
            (beg)
            (end))
       (when (search-forward-regexp drawer-beg-regexp bound t)
         (when inside
           (forward-line))
         (setq beg (line-beginning-position))
         (goto-char beg)
         (when (search-forward-regexp drawer-end-regexp bound t)
           (setq end (line-beginning-position))))
       (if (and (not (null beg))
                (not (null end)))
           (cons beg end)
         (when force
           (goto-char (cdr (org-get-property-block)))
           (forward-line 1)
           (open-line 1)
           (indent-for-tab-command)
           (insert ":" name ":\n")
           (indent-for-tab-command)
           (insert ":END:")
           (orgability--drawer-block name nil inside)))))))

(defun orgability--drawer-list-elements (name &optional parser prefix)
  "Get the content of NAME drawer as a list."
  (setq prefix (or prefix orgability--drawer-list-prefix))
  (setq parser (or parser #'identity))
  (let ((block (orgability--drawer-block name t t)))
    (orgability--with-entry
     (seq-map
      (lambda (line)
        (funcall parser (string-trim-left line prefix)))
      (split-string (buffer-substring-no-properties (car block) (cdr block)) "\n" t)))))

(defun orgability--drawer-link-url-parser (link)
  "Parse title from an org LINK."
  (replace-regexp-in-string "\\[\\[\\(.*\\)\\]\\[.*\\]\\]" "\\1" link))

(defun orgability--drawer-link-title-parser (link)
  "Parse title from an org LINK."
  (replace-regexp-in-string "\\[\\[.*\\]\\[\\(.*\\)\\]\\]" "\\1" link))

(defun orgability--drawer-link-parser (link)
  "Parse (url . title) from an org LINK."
  (cons
   (orgability--drawer-link-url-parser link)
   (orgability--drawer-link-title-parser link)))

(defun orgability--drawer-has-element (name el &optional testfn prefix)
  "Returns non-nil if NAME drawer contains element EL."
  (setq prefix (or prefix orgability--drawer-list-prefix))
  (setq testfn (or testfn #'orgability--link-p))
  (seq-contains (orgability--drawer-list-elements name nil prefix)
                el
                testfn))

(defun orgability--drawer-add-element (name el &optional testfn prefix)
  "Add element EL to NAME drawer."
  (setq prefix (or prefix orgability--drawer-list-prefix))
  (unless (orgability--drawer-has-element name el testfn prefix)
    (orgability--with-entry
     (goto-char (cdr (orgability--drawer-block name t)))
     (newline-and-indent)
     (forward-line -1)
     (insert (concat prefix el)))))

(defun orgability--drawer-del-element (name el &optional parser testfn prefix)
  (setq prefix (or prefix orgability--drawer-list-prefix))
  (setq parser (or parser #'identity))
  (setq testfn (or testfn #'orgability--link-p))
  (let ((block (orgability--drawer-block name t t)))
    (orgability--with-entry
     (goto-char (car block))
     (while (< (point) (cdr block))
       (if (funcall
            testfn
            el
            (funcall parser
                     (string-trim-left
                      (buffer-substring-no-properties
                       (line-beginning-position)
                       (line-end-position))
                      prefix)))
           (kill-whole-line)
         (forward-line 1)
         (beginning-of-line))))))

(defun orgability--link-p (url content)
  "Returns non-nil when CONTENT is a link to URL."
  (or (string-equal content url)
      (string-match-p (concat "\\[\\[" url "\\]\\[.*\\]\\]") content)))

(provide 'orgability-drawer)

;;; orgability-drawer.el ends here


orgability.el
;;; orgability.el --- reading list manager with offline access support.

;; Copyright (c) 2018 Boris Buliga

;; Author: Boris Buliga <boris@d12frosted.io>
;; Maintainer: Boris Buliga <boris@d12frosted.io>
;; Created: 25 Apr 2018

;; Keywords: org-mode
;; Homepage: https://github.com/d12frosted/orgability

;; Package-Version: 0.2.0
;; Package-Requires: ((org "9.1.0") (org-cliplink "0.2") (org-board "1136") (org-brain "0.5"))

;; This file is not part of GNU Emacs.
;;; License: GPLv3

;;; Commentary:
;;
;; Please checkout home page.

;;; Code:
;;

(require 'org-cliplink)
(require 'org-board)

(require 'orgability-utils)
(require 'orgability-drawer)
(require 'orgability-brain)

(defvar orgability-file (concat user-home-directory "orgability.org")
  "File for storing reading list.")

(defvar orgability-add-to-top t
  "If non-nil, entries are added to the top of the `orgability-file'.")

(defvar orgability-todo-keyword "TODO"
  "If non-nil, entries are added with a selected todo keyword.")

(defvar orgability-active-timestamp nil
  "If non-nil, ADDED timestamp is active.")

(defvar orgability-auto-archive t
  "If non-nil, entry is automatically archived.")

(defvar orgability-use-relative-archive-url t
  "If non-nil, use relative links to archive.

Useful when using several computers with different $HOME
directories.")

(defvar orgability-auto-agenda-redo t
  "If non-nil, redo agenda on clip.")

(defconst orgability-title "Reading list"
  "Title of `orgability-file'.")

(defconst orgability-category "reading-list"
  "Category of `orgability-file' entries.")

(defconst orgability-file-tags "reading"
  "Tags of `orgability-file' entries.")

(defconst orgability-topics-drawer "topics")

(defvar orgability-extract-http-title-f
  'org-cliplink-retrieve-title-synchronously
  "Function to extract title from http URL.")

(defun orgability-create-file (file)
  "Create an orgability FILE if it doesn't exist."
  (unless (file-exists-p file)
    (with-temp-file file
      (insert "#+TITLE: " orgability-title "\n"
              "#+CATEGORY: " orgability-category "\n"
              "#+FILETAGS: " orgability-file-tags "\n\n"))))

;;;###autoload
(defun orgability-open ()
  "Open archived file for entry at point."
  (interactive)
  (orgability--with-entry
   (call-interactively #'org-board-open)))

;;;###autoload
(defun orgability-clip ()
  "Store an URL from the clipboard."
  (interactive)
  (let ((url (substring-no-properties (current-kill 0))))
    (orgability-store-url url)))

;;;###autoload
(defun orgability-store-url (url)
  "Store an URL."
  (interactive "sUrl: ")
  (let ((title (funcall orgability-extract-http-title-f url)))
    (orgability-add-entry title url)))

(defun orgability-add-entry (title url &optional props)
  "Add read entry with TITLE and URL and optional PROPS."
  (unless title
    (setq title (read-string "Title: ")))
  (unless url
    (user-error "title is nil"))
  (unless orgability-file
    (user-error "`orgability-file' is not set"))
  (orgability-create-file orgability-file)
  (save-excursion
    (let ((buffer (find-file-noselect orgability-file)))
      (with-current-buffer buffer
        (if (and orgability-add-to-top
                 (progn (goto-char (point-min))
                        (search-forward "*" nil t)))
            (previous-line)
          (goto-char (point-max)))
        (org-insert-heading nil nil t)
        (when orgability-todo-keyword
          (insert orgability-todo-keyword " "))
        (insert title "\n")
        (org-set-property "URL" url)
        (org-set-property "ADDED" (concat
                                   (if orgability-active-timestamp "<" "[")
                                   (format-time-string "%Y-%02m-%02d")
                                   (if orgability-active-timestamp ">" "]")))
        (mapc (lambda (p) (org-set-property (car p) (cdr p))) props)
        (save-buffer)
        (when orgability-auto-archive
          (org-board-archive)
          (when orgability-use-relative-archive-url
            (let ((link (car (last
                              (org-entry-get-multivalued-property
                               (point) "ARCHIVED_AT")))))
              (org-set-property
               "ARCHIVED_AT"
               (replace-regexp-in-string
                (file-name-directory (buffer-file-name))
                ""
                link))))))
      (when org-agenda-buffer
        (with-current-buffer org-agenda-buffer
          (org-agenda-redo))))))

;;;###autoload
(defun orgability-add-topic ()
  "Add an `org-brain' topic to reading entry under point.

It works in a two-way fashion, meaning that the reading entry is
added as a resource to topic."
  (interactive)
  (orgability--with-entry
   (let* ((entry (orgability-brain-choose-entry))
          (link (orgability-brain-get-link entry))
          (id (org-id-get-create)))
     (unless (orgability--drawer-has-element orgability-topics-drawer link)
       ;; TODO: make sure that it's not being double added
       (orgability-brain-add-resource id entry)
       (orgability--drawer-add-element
        orgability-topics-drawer
        (org-make-link-string (orgability-brain-get-link entry)
                              (org-brain-title entry))))))
  (ignore-errors
    (org-agenda-redo)))

;;;###autoload
(defun orgability-delete-topic ()
  "Select and delete topic for the entry at point.

It works in a two-way fashion, meaning that the reading entry is
remove from resources of the topic."
  (interactive)
  (orgability--with-entry
   (let* ((id (org-id-get-create))
          (topics (orgability-list-topics))
          (target (completing-read
                   "Topic: "
                   (mapcar #'cdr topics)))
          (link (car (find-if (lambda (x)
                                (string-equal target
                                              (cdr x)))
                              topics))))
     ;; TODO: use orgability--drawer-del-element for this
     (orgability-brain-delete-resource id (orgability--unwrap-link link))
     (orgability--drawer-del-element orgability-topics-drawer link)))
  (ignore-errors
    (org-agenda-redo)))

(defun orgability-list-topics ()
  "Get the topics list of entry at point."
  (interactive)
  (orgability--drawer-list-elements
   orgability-topics-drawer
   #'orgability--drawer-link-parser))

(defvar orgability-agenda-topics-column 24
  "Width of topics block in `org-agenda'.")

(defun orgability-agenda-list-topics ()
  "Returns string with topics to be inserted to `org-agenda'."
  (let* ((topics (orgability-list-topics))
         (cl orgability-agenda-topics-column)
         (l (length (string-join (seq-map #'cdr topics) " ")))
         (extra-space (if (< l cl)
                          (make-string (- cl l) ? )
                        "")))
    (concat (string-join (seq-map (lambda (x)
                                    (org-make-link-string (car x) (cdr x)))
                                  topics)
                         " ")
            extra-space)))

(provide 'orgability)

;;; orgability.el ends here
